### PR TITLE
Prefs [TASK][28606] Create contract and presenter list restaurant

### PR DESCRIPTION
**Pull request by viet06061999**
## Related Tickets
- [Task#28606: Create contract and presenter listrestaurant](https://edu-redmine.sun-asterisk.vn/issues/28606)

## WHAT
Create contract and presenter list restaurant
## Review Checklist

Category | View Point | Description | Expected Reviewer Answer | Self review | Reviewer2 (name)
--- | --- | --- | --- | --- | ---
General |Code should be self-explanatory. | Get a feel of story reading, while going through the code. Use appropriate name for variables, functions and classes. If you are taking more time to understand the code, then either code needs refactoring or at least comments have to be written to make it clear. | YES |<li>- [x] yes</li>|<li>- [ ] yes</li>  
General |Is this function or class too big? If yes, is the function or class having too many responsibilities? | | NO |<li>- [x] no</li>|<li>- [ ] no</li>  
General | Does the code follow DRY?: Do not Repeat Yourself | Is there any redundant or duplicate code? DRY (Do not Repeat Yourself) principle: The same code should not be repeated more than twice. Consider reusable services, functions and components. Consider generic functions and classes. | YES |<li>- [x] yes</li>|<li>- [ ] yes</li>  
General |Exception handling and cleanup (dispose) resources | | YES |<li>- [x] yes</li>|<li>- [ ] yes</li>  
General | Does the code build and execute correctly? | | YES |<li>- [x] yes</li>|<li>- [ ] yes</li>  
Conventions | Does the source code have end-line errors? | | NO |<li>- [x] no</li>|<li>- [ ] no</li>  
Conventions | No hard coding, use constants/configuration values | | YES |<li>- [x] yes</li>|<li>- [ ] yes</li>  
Conventions | Does the code follow Sun* coding style and coding conventions? | https://github.com/framgia/coding-standards/tree/master/eng/android | YES |<li>- [x] yes</li>|<li>- [ ] yes</li>  
Redmine | Does the ticket follow Sun* Redmine working process?  | https://github.com/framgia/Training-Guideline/blob/master/WorkingProcess/redmine/redmine.md| YES |<li>- [x] yes</li>|<li>- [ ] yes</li>  
Documentation | Is there any incomplete code? If so, should it be removed or flagged with a suitable marker like ‘TODO’? |  | YES |<li>- [x] yes</li>|<li>- [ ] yes</li>  
## Notes (Optional)
*(Impacted Areas in Application(List features, api, models or services that this PR will affect))*

*(List gem, library third party add new)*

*(Other notes)*

### CODE REVIEW

File: app/src/main/java/com/sunasterisk/travelapp/ui/base/BaseMVPActivity.kt [hunk 1]
```diff
@@ -2,18 +2,27 @@ package com.sunasterisk.travelapp.ui.base
...
+    BaseActivity(), BaseContract.View<T> {
+
+    protected abstract val presenter: T
 
     private var progressDialog: Dialog? = null
 
+    @Suppress("UNCHECKED_CAST")
+    override fun onCreate(savedInstanceState: Bundle?) {
+        (this as? V)?.let { presenter.onAttach(it) }
+        super.onCreate(savedInstanceState)
```
Comment (aug16vcc) [inline app/src/main/java/com/sunasterisk/travelapp/ui/base/BaseMVPActivity.kt:23]:
```suggestion
        super.onCreate(savedInstanceState)
        (this as? V)?.let { presenter.attach(it) }
```

File: app/src/main/java/com/sunasterisk/travelapp/ui/list/restaurant/RestaurantListPresenter.kt [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+                override fun onSuccess(data: List<Restaurant>) {
+                    view?.updateRestaurants(data)
+                    view?.dismissProgressDialog()
+                }
+
+                override fun onError(throwable: Throwable) {
+                    view?.onError(throwable.message)
+                    view?.dismissProgressDialog()
+                }
+            })
```
Comment (aug16vcc) [inline app/src/main/java/com/sunasterisk/travelapp/ui/list/restaurant/RestaurantListPresenter.kt]: Gọi `dismissProgressDialog` trong `updateRestaurant` của view luôn. Sau này có thêm logic update hoặc error mà bị quên gọi dismiss thì sao?

**Review by aug16vcc — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/sunasterisk/travelapp/ui/base/BaseMVPActivity.kt (hunk 1 above)
Comment (viet06061999) [reply app/src/main/java/com/sunasterisk/travelapp/ui/base/BaseMVPActivity.kt:23]: em gọi super sau vì onCreate của super thực hiện hàm initComponent() nếu để super ở trước có thể khi init thực hiện  view của presenter vẫn null

File: app/src/main/java/com/sunasterisk/travelapp/ui/list/restaurant/RestaurantListPresenter.kt (hunk 2 above)
Comment (viet06061999) [reply app/src/main/java/com/sunasterisk/travelapp/ui/list/restaurant/RestaurantListPresenter.kt]: Em chưa hiểu ý của anh lắm, em show dialog trước khi request, sau khi xử lý reponse xong thì dissmiss thì ảnh hưởng gì đến logic của update và error ạ. Với cả gọi dissmiss trong update thì vẫn phải gọi trong onError chứ ạ

Comment (aug16vcc) [reply app/src/main/java/com/sunasterisk/travelapp/ui/list/restaurant/RestaurantListPresenter.kt]: Thì e để dismiss trong update với error luôn. E để đây nếu có thêm callback và e quên dismiss thì sao?

**Pull request merged**